cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -175,17 +175,23 @@ private ProgramStatus getExistingAppProgramStatus(ApplicationSpecification appSp
         }
         ensureAccess(programId);
 
-        if ((programId.getType() == ProgramType.MAPREDUCE || programId.getType() == ProgramType.SPARK) &&
-          !store.getRuns(programId, ProgramRunStatus.RUNNING, 0, Long.MAX_VALUE, 1).isEmpty()) {
-          // MapReduce program exists and running as a part of Workflow
+        boolean programTypeInsideWorkflow =
+          programId.getType() == ProgramType.MAPREDUCE || programId.getType() == ProgramType.SPARK;
+        boolean runningRunRecords = !store.getRuns(programId, ProgramRunStatus.RUNNING, 0, Long.MAX_VALUE, 1).isEmpty();
+        // If no runtime info exists for a program, the program is considered running if there is a RUNNING run record,
+        // or the program exists as part of a workflow and it is still in a STARTING state
+        if (runningRunRecords || (programTypeInsideWorkflow && !store.getRuns(programId, ProgramRunStatus.STARTING, 0,
+                                                                              Long.MAX_VALUE, 1).isEmpty())) {
           return ProgramStatus.RUNNING;
         }
         return ProgramStatus.STOPPED;
       }
       throw new IllegalStateException("Webapp status is not supported");
     }
 
-    return runtimeInfo.getController().getState().getProgramStatus();
+    String runId = runtimeInfo.getController().getRunId().getId();
```

Comment:
Its possible that the runtime info controller doesn't exist (right after program completion), but the COMPLETED run record hasn't been persisted yet, right?  (such as if the completed state message is still unprocessed in TMS)?
What would be the behavior of this method in that case?